auto-updater/paper_parser.py

```diff
@@ -0,0 +1,419 @@
+            file_path = Path(embeddings_subdir, filename)
+            with open(file_path) as ifh:
+                for line_num, line in enumerate(ifh):
+                    # Only copy header line from the first file
+                    if line_num == 0:
+                        if idx == 0:
+                            ofh.write(line)
+                        continue
+
+                    pmc_id = line.split('\t')[pmc_col]
+                    if pmc_id not in merged_pmc:
+                        ofh.write(line)
+                        merged_pmc.add(pmc_id)
+
+    # Set combined output file read-only
+    set_read_only(combined_embeddings_filename)
+
+    # Return the number of new papers found
+    return len(merged_pmc)
+
+
+def combine_new_token_counts(token_counts_subdir, combined_token_counts_filename):
+    """
+    Combine token_counts files generated by each process into a single one.
+    This is a little more complex, because each row's `document` column
+    in input files are NOT unique.
+    """
+
+    sub_files = sorted(os.listdir(token_counts_subdir))
+    merged_pmc = set()
+
+    with open(combined_token_counts_filename, 'w', newline='') as ofh:
+        fieldnames = ['document', 'lemma', 'count']
+        writer = csv.DictWriter(ofh, fieldnames=fieldnames, delimiter='\t')
+        writer.writeheader()
+
+        for filename in sub_files:
+            file_path = Path(token_counts_subdir, filename)
+            with open(file_path, newline='') as ifh:
+                prev_pmc = None
+                csv_reader = csv.DictReader(ifh, delimiter='\t')
+                for row in csv_reader:
+                    pmc_id = row['document']
+
+                    if pmc_id in merged_pmc:
+                        continue
+
+                    if pmc_id != prev_pmc: # enter a new token count session
+                        if prev_pmc:
+                            merged_pmc.add(prev_pmc)
+                        prev_pmc = pmc_id
+
+                    writer.writerow(row)
+
+    # Set combined output file read-only
+    set_read_only(combined_token_counts_filename)
+
+
+def parse_new_papers(
+        tarball_dir,
+        prev_pmc_list_filename,
+        word_model_vector_filename,
+        new_papers_dir,
+        new_pmc_list_basename,
+        new_embeddings_basename,
+        new_token_counts_basename,
+        parallel=4
+):
+    """Process tarball files and find new papers."""
+
+    all_filenames = os.listdir(tarball_dir)
+    tarball_files = [x for x in all_filenames if x.endswith(".xml.tar.gz")]
+
+    pmc_list_subdir = Path(new_papers_dir, 'pmc_list_sub')
+    os.makedirs(pmc_list_subdir, exist_ok=True)
+
+    embeddings_subdir = Path(new_papers_dir, 'embeddings_sub')
+    os.makedirs(embeddings_subdir, exist_ok=True)
+
+    token_counts_subdir = Path(new_papers_dir, 'token_counts_sub')
+    os.makedirs(token_counts_subdir, exist_ok=True)
+
+    pool = mp.Pool(parallel)
```

Comment:
For personal reference: is Pool better to use than multiprocess' Process class? I usually resort to the latter, which may or may not be a good idea.

Reply:
It depends. In the current situation, there's no difference between using a pool or launching processes manually, because there are only 8 tarball files to process.

But if you have a lot of jobs to run (say, more than 500), then using a pool will be much better, because the maximum number of sub-processes your process (no matter it's written in Python or any other languages) are limited by the OS. (Of course you can increase this limit at OS level, but that's another story). Using the pool like `Pool(n)` allows you to specify at most `n` sub-processes. When a job in the pool is done, another job will be assigned to the pool (kind of like producer-consumer pattern).